proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+  let other = x
+  let _ = move(x)
+  useX(move(other))
+  consumeX(other)
+  useX(x)
+}
+```
+
+we get separate diagnostics for each variable:
+
+```
+test.swift:7:15: error: 'x' used after being moved
+  let x = ...
+          ^
+test.swift:10:11: note: move here
+  let _ = move(x)
+          ^
+test.swift:13:3: note: use here
+  useX(x)
+  ^
+test.swift:9:7: error: 'other' used after being moved
+  let other = x
+      ^
+test.swift:11:8: note: move here
+  useX(move(other))
+       ^
+test.swift:12:3: note: use here
+  consumeX(other)
+  ^
+```
+
+If one applies move to a local `var`, then a new value can be assigned into
+the variable after an old value has been moved out. One can
+begin using the var again after one re-assigns to the var:
+
+```
+func f() {
+  var x = getValue()
+  let _ = move(x)
+  useX(x) // error: no value in x
+  x = getValue()
+  useX(x) // ok, x has a new value here
+}
+```
+
+This follows from move being applied to the binding (`x`), not the value in the
+binding (the value returned from `getValue()`).
+
+We also support applying the move operation to consuming function arguments:
+
+```
+func f(_ x: __owned SomeClassType) {
+    let _ = move(x)
+    useX(x) // !! Error! Use of x after move
+}
+```
+
+Normal arguments are passed by borrow, meaning that the lifetime of the value
+is managed by the caller. Although we could allow `move` on these arguments,
+shortening the syntactic lifetime of the variable, doing so would have no
+practical effect on the value's lifetime at runtime, so we choose to leave this
+disallowed for now.
+
+On the other hand, one can `move` out of an `inout`
+function argument. Like a `var`, the `inout` argument can be reassigned after
+being moved from and used again; however, since the final value of an
+`inout` argument is passed back to the caller, an `inout` argument *must* be
+reassigned by the callee before it returns:
+
+```
+func f(_ buffer: inout Buffer) { // error: 'buffer' not reinitialized after move!
+  let b = move(buffer)           // note: move was here
+  b.deinitialize()
+  ... write code ...
+}                                // note: return without reassigning inout argument `buffer`
+```
+
+But we can to re-initialize `buffer` by writing the following code:
+
+```
+func f(_ buffer: inout Buffer) {
+  let b = move(buffer)
+  b.deinitialize()
+  // ... write code ...
+  // We re-initialized buffer before end of function so the checker is satisfied
+  buffer = getNewInstance()
+}
+```
+
+Move analysis understands that `defer` statements run before every scope exit,
+so `defer` can also be used to reinitialize an `inout` or `var` after a move.
+So we can also write the above as:
+
+```
+func f(_ buffer: inout Buffer) {
+  let b = move(buffer)
+  // Ensure the buffer is reinitialized before we exit.
+  defer { buffer = getNewInstance() }
+  b.deinitialize()
+  // ... write code ...
+}
+```
+
+
+In the future, we may add support for globals and stored properties, although
```

Comment:
Again, you're talking about the implementation.  You need to explain why you do not think the feature requires this support today.

Hmmm....  This paragraph might be better off moved to "alternatives considered".